test/hotspot/jtreg/compiler/c2/TestFloat16Reduction.java

```diff
@@ -0,0 +1,166 @@
+            res = Float.floatToFloat16(Float.float16ToFloat(res) - Float.float16ToFloat(arr[i]));
+        }
+        return (long)res;
+    }
+
+    @Check(test="SUBReduceLong")
+    void checkSUBReduceLong(long actual) {
+        Verify.checkEQ(actual, GOLDEN_SUB);
+    }
+
+    @Test
+    @IR(counts = {IRNode.MUL_HF, " >0 "}, applyIfCPUFeature = {"avx512_fp16", "true"})
+    @IR(counts = {IRNode.MUL_HF, " >0 "}, applyIfCPUFeatureAnd = {"fphp", "true", "asimdhp", "true"})
+    long MULReduceLong() {
+        short res = 0;
+        for (int i = 0; i < arr.length; i++) {
+            res = Float.floatToFloat16(Float.float16ToFloat(res) * Float.float16ToFloat(arr[i]));
+        }
+        return (long)res;
+    }
+
+    @Check(test="MULReduceLong")
+    void checkMULReduceLong(long actual) {
+        Verify.checkEQ(actual, GOLDEN_MUL);
+    }
+
+    @Test
+    @IR(counts = {IRNode.DIV_HF, " >0 "}, applyIfCPUFeature = {"avx512_fp16", "true"})
+    @IR(counts = {IRNode.DIV_HF, " >0 "}, applyIfCPUFeatureAnd = {"fphp", "true", "asimdhp", "true"})
+    long DIVReduceLong() {
+        short res = 0;
+        for (int i = 0; i < arr.length; i++) {
+            res = Float.floatToFloat16(Float.float16ToFloat(res) / Float.float16ToFloat(arr[i]));
+        }
+        return (long)res;
+    }
+
+    @Check(test="DIVReduceLong")
+    void checkDIVReduceLong(long actual) {
+        Verify.checkEQ(actual, GOLDEN_DIV);
+    }
+
+    @Test
+    @IR(counts = {IRNode.MAX_HF, " >0 "}, applyIfCPUFeature = {"avx512_fp16", "true"})
+    @IR(counts = {IRNode.MAX_HF, " >0 "}, applyIfCPUFeatureAnd = {"fphp", "true", "asimdhp", "true"})
+    long MAXReduceLong() {
+        short res = 0;
+        for (int i = 0; i < arr.length; i++) {
+            res = Float.floatToFloat16(Math.max(Float.float16ToFloat(res), Float.float16ToFloat(arr[i])));
+        }
+        return (long)res;
+    }
+
+    @Check(test="MAXReduceLong")
+    void checkMAXReduceLong(long actual) {
+        Verify.checkEQ(actual, GOLDEN_MAX);
+    }
+
+    @Test
+    @IR(counts = {IRNode.MIN_HF, " >0 "}, applyIfCPUFeature = {"avx512_fp16", "true"})
+    @IR(counts = {IRNode.MIN_HF, " >0 "}, applyIfCPUFeatureAnd = {"fphp", "true", "asimdhp", "true"})
+    long MINReduceLong() {
+        short res = 0;
+        for (int i = 0; i < arr.length; i++) {
+            res = Float.floatToFloat16(Math.min(Float.float16ToFloat(res), Float.float16ToFloat(arr[i])));
+        }
+        return (long)res;
+    }
+
+    @Check(test = "MINReduceLong")
+    void checkMINReduceLong(long actual) {
+        Verify.checkEQ(actual, GOLDEN_MIN);
+    }
+
+    public TestFloat16Reduction() {
+        arr = new short[32];
+        G.fill(genHF, arr);
+
+        GOLDEN_ADD = ADDReduceLong();
+        GOLDEN_SUB = SUBReduceLong();
+        GOLDEN_MUL = MULReduceLong();
+        GOLDEN_DIV = DIVReduceLong();
+        GOLDEN_MAX = MAXReduceLong();
+        GOLDEN_MIN = MINReduceLong();
+    }
```

Comment:
A total nit, and optional: you could make the fields static, and just assign values as you declare the fields. That would save you doing it all in the constructor.